core/src/validator.rs

```diff
@@ -698,14 +702,38 @@ fn backup_and_clear_blockstore(ledger_path: &Path, start_slot: Slot, shred_versi
     drop(blockstore);
 }
 
+// Return true on error, indicating the validator should exit.
 fn wait_for_supermajority(
     config: &ValidatorConfig,
     bank: &Bank,
     cluster_info: &ClusterInfo,
     rpc_override_health_check: Arc<AtomicBool>,
-) {
-    if config.wait_for_supermajority != Some(bank.slot()) {
-        return;
+) -> bool {
+    if let Some(wait_for_supermajority) = config.wait_for_supermajority {
+        match wait_for_supermajority.cmp(&bank.slot()) {
+            std::cmp::Ordering::Less => return false,
+            std::cmp::Ordering::Greater => {
+                error!("Ledger does not have enough data to wait for supermajority, please enable snapshot fetch. Has {} needs {}", bank.slot(), wait_for_supermajority);
+                return true;
+            }
+            _ => {}
+        }
+    } else {
+        return false;
+    }
+
+    if let Some(expected_bank_hash) = config.expected_bank_hash {
+        if bank.hash() != expected_bank_hash {
+            error!(
+                "Bank hash({}) does not match expected value: {}",
+                bank.hash(),
+                expected_bank_hash
+            );
+            return true;
+        }
+    } else {
+        error!("Bank hash should be provided when waiting for supermajority slot");
```

Comment:
You can let `clap` handle this dependency for you by adding `.requires("expected_bank_hash")` to the `wait_for_supermajority` arg declaration

Reply:
ok, I'll add that in a follow-up

Reply:
Added it in this one actually.